[Prittam Ravi] iP

prit3010 opened the pull request:
# Duke

> "More than 5" - [Kevin Chang 💀](https://github.com/kevinchangjk) 

Amazing Task Manger needed?

Introducing DukePRO, the omnipotent task manager.

## Capabilities

Duke is perhaps the greatest innovation since the [Internet]

- Keeping track of your tasks
- Keeping track of deadlines
- Storing Tasks
- Deleting Tasks

## Setup

The installation and setup is easy. All you need to do is this:
1. Download the latest jar file from [here](https://github.com/prit3010/ip/releases)
2. Run the jar file with the following command
    bash
    java -jar /path/to/jar/file
    
3. You're all ready to use Duke! 🔥 

## Roadmap

- [X] Basic Task Manager Functions
- [ ] Generative Chatting Functionality
- [ ] AI Chatbot

## Contributing

Open Source Code. Help me to make this good

_Excerpt from Duke source code_

```
java
public class Duke {
    /**
     * Initializes and runs an instance of Duke.
     *
     */
    public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
}
```

Commit:
Add Gradle support

prit3010 committed:
Level-1

prit3010 committed:
Level-2

prit3010 committed:
Level-3

prit3010 committed:
Level-4

prit3010 committed:
A-TextUiTesting

prit3010 committed:
handle exceptions

prit3010 committed:
added delete feature

prit3010 committed:
Add saving functionality

prit3010 committed:
Add DateTime functionality

prit3010 committed:
Merge branch 'branch-Level-8'

prit3010 committed:
Refactor more OOP

prit3010 committed:
Organize into Duke Package

prit3010 committed:
Merge branch 'add-gradle-support'

prit3010 committed:
Add JUnit test cases

prit3010 committed:
Fix load data

prit3010 committed:
Fix build.gradle

prit3010 committed:
Add Java Docs

prit3010 committed:
Follow coding standard

prit3010 committed:
Add find functionality

prit3010 committed:
Change package name

prit3010 committed:
Merge branch 'branch-A-JavaDoc'

prit3010 committed:
Fix error

prit3010 committed:
Merge branch 'branch-A-CodingStandard'

prit3010 committed:
Merge branch 'branch-Level-9'

prit3010 committed:
Fix error

sjoann commented on src/main/java/duke/DateTimeConverter.java:
```diff
@@ -10,12 +10,26 @@ public class DateTimeConverter {
     protected static final DateTimeFormatter dateFormatPrint = DateTimeFormatter
```
I think it would be better if the constant names are in all uppercase

sjoann commented on src/main/java/duke/Ui.java:
```diff
@@ -5,28 +5,51 @@
+     * Outputs the greeting for Duke
+     *
+     */
     public static void intro() {
```
maybe you can use \n like what you did for other methods to shorten the code

sjoann commented on src/main/java/duke/Deadline.java:
```diff
@@ -6,17 +6,25 @@ public class Deadline extends Task {
         this.by = DateTimeConverter.formatDate(by);
     }
 
+    /** @inheritdoc */
```
I like that you use this annotation for the overridden methods :)

sjoann reviewed (commented):
Overall great work ! the code is easy to read and there are also JavaDocs for the methods

prit3010 committed:
Add GUI functionality

PokezardVGC commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public String changeFormat() {
+        String done = isDone ? "1" : "0";
+        return "E | " + done + " | " + this.description + " | " + this.at;
```
Would string formatter look neater for this?

PokezardVGC commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,31 @@
+            stage.setScene(scene);
+            fxmlLoader.<MainWindow>getController().setDuke(duke);
+            stage.show();
+        } catch (IOException e) {
```
Maybe can specify the exception instead of catching all exceptions?

PokezardVGC commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     */
+    public String process(String cmd) {
+        try {
+            if (cmd.toLowerCase().equals("bye")) {
```
Believe you can extract out the first input and lowercase all of it at once instead of repeating the code

optionalemon commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+    /** @inheritdoc */
+    @Override
+    public String changeFormat() {
+        String done = isDone ? "1" : "0";
```
Any reason of doing this rather than changing isDone and use the usual toString function instead?

optionalemon commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+
+    @Override
+    public String toString() {
+        return "\uD83D\uDE14 OPPS!";
```
I like how an emoji is being used here!

optionalemon commented on src/main/java/duke/OutOfListException.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
+
+public class OutOfListException extends DukeException {
```
I like how you specify different kind of exception as child of DukeException (I merely use DukeException itself), it is much neater!

optionalemon commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+        }
+    }
+
+    public String handleUnmark(String cmd) throws DukeException {
```
Maybe include more javaDoc in this file? :)

optionalemon commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     */
+    public static String intro() {
+        String res = UNDERLINE + "\n" + "Hello! I'm Duke!"  + "\n" + "What can I do for you?" + "\n" + UNDERLINE;
```
Maybe for the repetitive UNDERLINE used in this file, you can use a function to format all the output with that?

optionalemon commented on src/test/java/duke/ToDoTest.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @Test
+    public void testChangeFormat() {
```
Underscore can be used in test methods name! e.g. featureUnderTest_testScenario_expectedBehavior() as mentioned in the java coding practice website, perhaps you can try using that format?

prit3010 committed:
Improve Code Quality

Change parts of the code so that it upholds a better standard of code
quality. Extract out repeated parts of the code so as to adhere to the
DRY principle.

prit3010 committed:
Use assertions to check important assumptions

There are various section of the codes where it is assumed that the
input is not null. In some corner cases, these might cause the program
to fail.

Using assertions define assumptions about the program state so that
the runtime can verify them. These potentially help catch bugs.

prit3010 committed:
Merge pull request #1 from prit3010/branch-A-Assertation

Use assertions to check important assumptions

prit3010 committed:
Merge pull request #2 from prit3010/branch-A-CodeQuality

Improve Code Quality

prit3010 committed:
Add Duplicate Functionality

Users may forget that they added a task already and try to add it again
which would mean that there is duplicate tasks within the list

Add the duplicate functionality so that it throws an error if there is
a task with the same task description already in the list.

prit3010 committed:
Update User Guide

prit3010 committed:
Update GUI

prit3010 committed:
Add User Guide

prit3010 committed:
Add Ui image

prit3010 committed:
Edit User Guide

prit3010 committed:
Fix build file
